model/modelprocessor/datastream.go

```diff
@@ -47,8 +47,13 @@ func (s *SetDataStream) setDataStream(event *model.APMEvent) {
 		event.DataStream.Type = datastreams.LogsType
 		event.DataStream.Dataset = model.ErrorsDataset
 	case model.LogProcessor:
-		event.DataStream.Type = datastreams.LogsType
-		event.DataStream.Dataset = model.AppLogsDataset
+		if event.Firehose != nil {
+			event.DataStream.Type = datastreams.LogsType
+			event.DataStream.Dataset = model.FirehoseDataset
+		} else {
+			event.DataStream.Type = datastreams.LogsType
+			event.DataStream.Dataset = model.AppLogsDataset
+		}
```

Comment:
I'd rather not proliferate logic in here. Can you please make a change to this processor so that it leaves Type and Dataset alone if they are already set? In that case it should only set Namespace. Then the Firehose handler can explicitly set the Type and Dataset fields.